static/js/message_fetch.js

```diff
@@ -128,14 +128,22 @@ exports.load_messages = function (opts) {
 };
 
 exports.load_messages_for_narrow = function (opts) {
+    var num_before = consts.narrow_before;
+    var num_after = consts.narrow_after;
+
     message_fetch.load_messages({
         anchor: opts.then_select_id.toFixed(),
-        num_before: consts.narrow_before,
-        num_after: consts.narrow_after,
+        num_before: num_before,
+        num_after: num_after,
         msg_list: message_list.narrowed,
         use_first_unread_anchor: opts.use_initial_narrow_pointer,
-        cont: function () {
+        cont: function (messages) {
             exports.reset_for_new_narrow();
+            // Err on the side of continuing to fetch.  But for
+            // small narrows we can short circuit fetching early.
+            if (messages.length < num_before + num_before - 2) {
+                fetch_status.finish_fetching();
+            }
```

Comment:
I think this logic isn't correct; I think you want `min(num_before, num_after) - 1` for this algorithm.

Regardless, we can do better than this; I believe that the `load_messages` API endpoint returns an `anchor` if we didn't specify one (i..e if we used `use_first_unread_anchor`), so we always have the anchor in the return code path.  We should be able to thus compare the number of messages with an ID above or below the narrow that were returned when narrowing.  

(Or we could just do that message ID counting logistics in the backend and send the result, and make life a little easier for clients, which might be an even better idea).